packages/realm-server/handlers/handle-publish-realm.ts

```diff
@@ -0,0 +1,270 @@
+          | 'owner_username'
+          | 'source_realm_url'
+          | 'published_realm_url'
+          | 'last_published_at'
+        >[];
+        publishedRealmData = results[0];
+        realmUsername = `realm/${PUBLISHED_DIRECTORY_NAME}_${publishedRealmData.id}`;
+
+        let lastPublishedAt = new Date().toISOString();
+        await query(dbAdapter, [
+          `UPDATE published_realms SET last_published_at =`,
+          param(lastPublishedAt),
+          `WHERE published_realm_url =`,
+          param(publishedRealmURL),
+        ]);
+        publishedRealmData.last_published_at = lastPublishedAt;
+      } else {
+        let publishedRealmId = uuidv4();
+        realmUsername = `realm/${PUBLISHED_DIRECTORY_NAME}_${publishedRealmId}`;
+        let { valueExpressions, nameExpressions } = asExpressions({
+          id: publishedRealmId,
+          owner_username: realmUsername,
+          source_realm_url: sourceRealmURL,
+          published_realm_url: publishedRealmURL,
+          last_published_at: new Date(),
+        });
+
+        let results = (await query(
+          dbAdapter,
+          insert('published_realms', nameExpressions, valueExpressions),
+        )) as Pick<
+          PublishedRealmTable,
+          | 'id'
+          | 'owner_username'
+          | 'source_realm_url'
+          | 'published_realm_url'
+          | 'last_published_at'
+        >[];
+        publishedRealmData = results[0];
+
+        let { userId: newUserId } = await registerUser({
+          matrixURL: matrixClient.matrixURL,
+          displayname: realmUsername,
+          username: realmUsername,
+          password: await passwordFromSeed(realmUsername, realmSecretSeed),
+          registrationSecret: await getMatrixRegistrationSecret(),
+        });
+        userId = newUserId;
+
+        await insertPermissions(dbAdapter, new URL(publishedRealmURL), {
+          [userId]: DEFAULT_PERMISSIONS,
+          [ownerUserId]: DEFAULT_PERMISSIONS,
+          '*': ['read'],
+        });
+      }
+      log.debug(
+        `created realm bot user '${userId}' for new realm ${publishedRealmURL}`,
+      );
+
+      let pathNameParts = new URL(sourceRealmURL).pathname
+        .split('/')
+        .filter((p) => p);
+      if (pathNameParts.length < 1) {
+        throw new Error('Could not determine source realm folder');
+      }
+      let sourceRealmPath = resolve(join(realmsRootPath, ...pathNameParts));
+      let publishedDir = join(realmsRootPath, PUBLISHED_DIRECTORY_NAME);
+      let publishedRealmPath = join(publishedDir, publishedRealmData.id);
+      copySync(sourceRealmPath, publishedRealmPath);
+      ensureDirSync(publishedRealmPath);
+
+      if (existingPublishedRealm) {
+        virtualNetwork.unmount(existingPublishedRealm.handle);
+      }
+      let realm = createAndMountRealm(
+        publishedRealmPath,
+        publishedRealmURL,
+        realmUsername,
+        new URL(sourceRealmURL),
+      );
+      await realm.start();
```

Comment:
This is process-specific, and we have multiple processes running (workers, worker manager, potentially multiple realm server instances). Seems problematic.

Reply:
Do you mean that if an existing published realm is republished, we don’t need to re-create and re-start the realm?